File: config-model/src/test/scala/com/yahoo/vespa/model/container/jersey/xml/RestApiTest.scala
// Copyright 2016 Yahoo Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vespa.model.container.jersey.xml

import scala.language.implicitConversions
import RestApiTest._
import scala.xml.Elem
import scala.collection.JavaConverters._
import com.yahoo.config.model.builder.xml.test.DomBuilderTest
import org.w3c.dom.Element
import org.junit.{Ignore, Test}
import org.junit.Assert._
import com.yahoo.vespa.model.container.component.{Component, Handler}
import com.yahoo.vespa.model.container.jersey.{JerseyHandler => ModelJerseyHandler, RestApi, Jersey2Servlet, RestApiContext}
import com.yahoo.vespa.model.container.xml.ContainerModelBuilderTestBase
import com.yahoo.vespa.model.container.xml.ContainerModelBuilderTestBase._
import com.yahoo.container.ComponentsConfig
import org.hamcrest.CoreMatchers.{is, nullValue, notNullValue, not, containsString, hasItem, hasItems}
import org.hamcrest.Matchers.{empty, contains}
import com.yahoo.container.jdisc.JdiscBindingsConfig
import com.yahoo.container.di.config.{JerseyInjectionConfig, JerseyBundlesConfig}
import com.yahoo.container.config.jersey.JerseyInitConfig
import com.yahoo.component.ComponentId

/**
 * @author gjoranv
 * @since 5.5
 */
class RestApiTest extends ContainerModelBuilderTestBase {

  trait TestApp {
    createModel(root, restApiXml)
    root.validate()
    getContainerCluster(ClusterId).prepare()

    val restApi = getContainerCluster(ClusterId).getRestApiMap.values().iterator().next()
    val handler = getContainerComponentNested(ClusterId, HandlerId).asInstanceOf[ModelJerseyHandler]
    val context = restApi.getContext
  }

  @Test
  def jersey_handler_has_correct_bindings() {
    new TestApp {
      assertThat(handler, not(nullValue()))
      assertThat(handler.getServerBindings, hasItems(HttpBinding, HttpsBinding))
    }
  }

  @Test
  def jersey_bindings_are_included_in_config() {
    new TestApp {
      val config = root.getConfig(classOf[JdiscBindingsConfig], ClusterId)
      assertThat(config.handlers(HandlerId).serverBindings(), hasItems(HttpBinding, HttpsBinding))
    }
  }

  @Test
  def jersey_handler_has_correct_bundle_spec() {
    new TestApp {
      assertThat(handler.model.bundleInstantiationSpec.bundle.stringValue(), is(ModelJerseyHandler.BUNDLE))
    }
  }

  @Test
  def config_has_correct_jersey_mapping() {
    new TestApp {
      val config = root.getConfig(classOf[JerseyInitConfig], handler.getConfigId)
      assertThat(config.jerseyMapping, is(Path))
    }
  }

  @Test
  def resource_bundles_are_included_in_config() {
    new TestApp {
      val config = root.getConfig(classOf[JerseyBundlesConfig], context.getConfigId)

      assertThat(config.bundles.size, is(1))
      assertThat(config.bundles(0).spec, is("my-jersey-bundle:1.0"))
    }
  }

  @Test
  def packages_to_scan_are_included_in_config() {
    new TestApp {
      val config = root.getConfig(classOf[JerseyBundlesConfig], context.getConfigId)
      assertThat(config.bundles(0).packages, contains("com.yahoo.foo"))
    }
  }

  @Test
  def jersey_handler_is_included_in_components_config() {
    new TestApp {
      val config = root.getConfig(classOf[ComponentsConfig], ClusterId)
      assertThat(config.toString, containsString(".id \"" + HandlerId + "\""))
    }
  }

  @Test
  def restApiContext_is_included_in_components_config() {
    new TestApp {
      val config = root.getConfig(classOf[ComponentsConfig], ClusterId)
      assertThat(config.toString, containsString(".id \"" + RestApiContextId + "\""))
    }
  }

  @Test
  def all_non_restApi_components_are_injected_to_RestApiContext() {
    new TestApp {
      def restApiContextConfig(config: ComponentsConfig) =
        (for {
          component <- config.components.asScala
          if component.classId == RestApiContext.CONTAINER_CLASS
        } yield component).head

      val componentsConfig = root.getConfig(classOf[ComponentsConfig], ClusterId)

      val clusterChildrenComponentIds = getContainerCluster(ClusterId).getAllComponents.asScala.map(
        child => child.getComponentId).toSet

      val restApiChildrenComponentIds = restApi.getChildren.values.asScala.map(
        child => child.asInstanceOf[Component[_, _]].getComponentId).toSet

      //TODO: Review: replace with filtering against RestApiContext.isCycleGeneratingComponent
      val cycleInducingComponents = Set("com.yahoo.container.handler.observability.ApplicationStatusHandler") map ComponentId.fromString

      val expectedInjectedConfigIds = clusterChildrenComponentIds -- restApiChildrenComponentIds -- cycleInducingComponents

      val injectedConfigIds = restApiContextConfig(componentsConfig).inject.asScala.map(
        inject => ComponentId.fromString(inject.id()))

      // Verify that the two sets are equal. Split in two asserts to get decent failure messages.
      assertThat("Not all required components are injected",
                 (expectedInjectedConfigIds -- injectedConfigIds).asJavaCollection, empty[Any])
      assertThat("We inject some components that should not be injected",
                 (injectedConfigIds -- expectedInjectedConfigIds).asJavaCollection, empty[Any])
    }
  }

  @Ignore // TODO: use for naming components instead
  @Test
  def jdisc_components_can_be_injected() {
    new TestApp {
      val config = root.getConfig(classOf[JerseyInjectionConfig], context.getConfigId)
      assertThat(config.inject(0).instance(), is("injectedHandler"))
      assertThat(config.inject(0).forClass(), is("com.yahoo.handler.Handler"))
    }
  }

  @Ignore // TODO: use for naming a non-existent component instead
  @Test(expected = classOf[IllegalArgumentException])
  def injecting_non_existent_component() {
    val restApiXml =
      <container version="1.0" id={ClusterId}>
        <rest-api path={Path}>
          <components bundle="my-jersey-bundle:1.0" />
          <inject jdisc-component="non-existent" for-class="foo" />
        </rest-api>

      </container>

    createModel(root, restApiXml)
    root.validate()

  }

  @Test
  def legacy_syntax_should_produce_valid_model() {
    val legacyXml =
      <container version="1.0" >
        <handler id={ModelJerseyHandler.CLASS}>
          <binding>{HttpBinding}</binding>
          <config name="jdisc.jersey.jersey-handler">
            <jerseyMapping>jersey</jerseyMapping>
          </config>
        </handler>
      </container>

    createModel(root, legacyXml)

    val handler = getContainerComponent("container", ModelJerseyHandler.CLASS).asInstanceOf[Handler[_]]
    assertThat(handler, not(nullValue()))
    assertThat(handler.getServerBindings, hasItem(HttpBinding))

    val bindingsConfig = root.getConfig(classOf[JdiscBindingsConfig], ClusterId)
    assertThat(bindingsConfig.handlers(ModelJerseyHandler.CLASS).serverBindings(), hasItem(HttpBinding))
  }

}

object RestApiTest {
  val Path = "rest/api"
  val HttpBinding = "http://*/" + Path + "/*"
  val HttpsBinding = "https://*/" + Path + "/*"
  val HandlerId = ModelJerseyHandler.CLASS + "-" + RestApi.idFromPath(Path)
  val RestApiContextId = RestApiContext.CONTAINER_CLASS + "-" + RestApi.idFromPath(Path)
  val InjectedComponentId = "injectedHandler"

  val ClusterId = "container"

  val restApiXml =
    <container version="1.0" id={ClusterId} jetty="true">
      <rest-api path={Path}>
        <components bundle="my-jersey-bundle:1.0">
          <package>com.yahoo.foo</package>
        </components>
      </rest-api>

      <handler id={InjectedComponentId} />
    </container>

  implicit def toDomElement(elem: Elem): Element = {
    DomBuilderTest.parse(elem.toString())
  }

}

File: config-model/src/test/java/com/yahoo/vespa/model/container/jersey/xml/RestApiTest.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vespa.model.container.jersey.xml;

import com.yahoo.component.ComponentId;
import com.yahoo.config.model.test.TestUtil;
import com.yahoo.container.ComponentsConfig;
import com.yahoo.container.di.config.JerseyBundlesConfig;
import com.yahoo.jdisc.http.ServletPathsConfig;
import com.yahoo.vespa.model.container.component.Component;
import com.yahoo.vespa.model.container.jersey.Jersey2Servlet;
import com.yahoo.vespa.model.container.jersey.RestApi;
import com.yahoo.vespa.model.container.jersey.RestApiContext;
import com.yahoo.vespa.model.container.xml.ContainerModelBuilderTestBase;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.w3c.dom.Element;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.not;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.containsInAnyOrder;

/**
 * @author gjoranv
 * @author bjorncs
 */
@Ignore // TODO: remove test
public class RestApiTest extends ContainerModelBuilderTestBase {
    private static final String PATH = "rest/api";
    private static final String REST_API_CONTEXT_ID = RestApiContext.CONTAINER_CLASS + "-" + RestApi.idFromPath(PATH);
    private static final String INJECTED_COMPONENT_ID = "injectedHandler";
    private static final String CLUSTER_ID = "container";

    private static final Element restApiXml = TestUtil.parse(
            "<container version=\"1.0\" id=\"" + CLUSTER_ID + "\">",
            "  <rest-api path=\"" + PATH + "\">",
            "    <components bundle=\"my-jersey-bundle:1.0\">",
            "      <package>com.yahoo.foo</package>",
            "    </components>",
            "  </rest-api>",
            "  <handler id=\"" + INJECTED_COMPONENT_ID + "\" />",
            "</container>");

    private RestApi restApi;
    private Jersey2Servlet servlet;
    private RestApiContext context;

    @Before
    public void setup() throws Exception {
        createModel(root, restApiXml);
        root.validate();
        getContainerCluster(CLUSTER_ID).prepare(root.getDeployState());
        restApi = getContainerCluster(CLUSTER_ID).getRestApiMap().values().iterator().next();
        servlet = restApi.getJersey2Servlet();
        context = restApi.getContext();
    }

    @Test
    public void jersey2_servlet_has_correct_binding_path() {
        assertThat(servlet, not(nullValue()));
        assertThat(servlet.bindingPath, is(PATH + "/*"));
    }

    @Test
    public void jersey2_servlet_has_correct_bundle_spec() {
        assertThat(servlet.model.bundleInstantiationSpec.bundle.stringValue(), is(Jersey2Servlet.BUNDLE));
    }

    @Test
    public void rest_api_path_is_included_in_servlet_config() {
        ServletPathsConfig config = root.getConfig(ServletPathsConfig.class, servlet.getConfigId());
        assertThat(config.servlets(servlet.getComponentId().stringValue()).path(), is(PATH + "/*"));
    }

    @Test
    public void resource_bundles_are_included_in_config() {
        JerseyBundlesConfig config = root.getConfig(JerseyBundlesConfig.class, context.getConfigId());
        assertThat(config.bundles().size(), is(1));
        assertThat(config.bundles(0).spec(), is("my-jersey-bundle:1.0"));
    }

    @Test
    public void packages_to_scan_are_included_in_config() {
        JerseyBundlesConfig config = root.getConfig(JerseyBundlesConfig.class, context.getConfigId());
        assertThat(config.bundles(0).packages(), contains("com.yahoo.foo"));
    }

    @Test
    public void jersey2_servlet_is_included_in_components_config() {
        ComponentsConfig config = root.getConfig(ComponentsConfig.class, CLUSTER_ID);
        assertThat(config.toString(), containsString(".id \"" + servlet.getComponentId().stringValue() + "\""));
    }

    @Test
    public void restApiContext_is_included_in_components_config() {
        ComponentsConfig config = root.getConfig(ComponentsConfig.class, CLUSTER_ID);
        assertThat(config.toString(), containsString(".id \"" + REST_API_CONTEXT_ID + "\""));
    }

    @Test
    public void all_non_restApi_components_are_injected_to_RestApiContext() {
        ComponentsConfig componentsConfig = root.getConfig(ComponentsConfig.class, CLUSTER_ID);

        Set<ComponentId> clusterChildrenComponentIds = getContainerCluster(CLUSTER_ID).getAllComponents().stream()
                .map(Component::getComponentId)
                .collect(Collectors.toSet());

        Set<ComponentId> restApiChildrenComponentIds = restApi.getChildren().values().stream()
                .map(child -> ((Component<?, ?>) child).getComponentId())
                .collect(Collectors.toSet());

        //TODO: try replacing with filtering against RestApiContext.isCycleGeneratingComponent
        ComponentId cycleInducingComponents = ComponentId.fromString("com.yahoo.container.handler.observability.ApplicationStatusHandler");

        Set<ComponentId> expectedInjectedConfigIds = new HashSet<>(clusterChildrenComponentIds);
        expectedInjectedConfigIds.removeAll(restApiChildrenComponentIds);
        expectedInjectedConfigIds.remove(cycleInducingComponents);

        Set<ComponentId> injectedConfigIds = restApiContextConfig(componentsConfig).inject().stream()
                .map(inject -> ComponentId.fromString(inject.id()))
                .collect(Collectors.toSet());

        // Verify that the two sets are equal. Split in two asserts to get decent failure messages.
        assertThat(
                "Not all required components are injected",
                injectedConfigIds,
                containsInAnyOrder(expectedInjectedConfigIds.toArray()));
        assertThat(
                "We inject some components that should not be injected",
                expectedInjectedConfigIds,
                containsInAnyOrder(injectedConfigIds.toArray()));
    }

    private static ComponentsConfig.Components restApiContextConfig(ComponentsConfig config) {
        return config.components().stream()
                .filter(component -> component.classId().equals(RestApiContext.CONTAINER_CLASS))
                .findFirst()
                .get();
    }

}
